rubix-core/src/main/java/com/qubole/rubix/core/RemoteReadRequestChain.java

```diff
@@ -61,78 +81,76 @@ public Integer call()
-                log.debug(String.format("PrefixLength: %d SuffixLength: %d", prefixBufferLength, suffixBufferLength));
-
-                // TODO: use single byte buffer for all three streams
-                /* TODO: also GC cost can be lowered by shared buffer pool, a small one.
-                    IOUtils.copyLarge method. A single 4kB byte buffer can be used to copy whole file
-
-                  */
-                if (prefixBufferLength > 0) {
-                    byte[] prefixBuffer = new byte[prefixBufferLength];
-                    log.debug(String.format("Trying to Read %d bytes into prefix buffer", prefixBufferLength));
-                    totalPrefixRead += readAndCopy(prefixBuffer, 0, mbuf, prefixBufferLength);
-                    log.debug(String.format("Read %d bytes into prefix buffer", prefixBufferLength));
-                }
-                log.debug(String.format("Trying to Read %d bytes into destination buffer", readRequest.getActualReadLength()));
-                int readBytes = readAndCopy(readRequest.getDestBuffer(), readRequest.destBufferOffset, mbuf, readRequest.getActualReadLength());
-                totalRequestedRead += readBytes;
-                log.debug(String.format("Read %d bytes into destination buffer", readBytes));
-                if (suffixBufferLength > 0) {
-                    // If already in reading actually required data we get a eof, then there should not have been a suffix request
-                    checkState(readBytes == readRequest.getActualReadLength(), "Acutal read less than required, still requested for suffix");
-                    byte[] suffixBuffer = new byte[suffixBufferLength];
-                    log.debug(String.format("Trying to Read %d bytes into suffix buffer", suffixBufferLength));
-                    totalSuffixRead += readAndCopy(suffixBuffer, 0, mbuf, suffixBufferLength);
-                    log.debug(String.format("Read %d bytes into suffix buffer", suffixBufferLength));
-                }
-            }
-        }
-        finally {
-            if (fc != null) {
-                fc.close();
+            int prefixBufferLength = (int) (readRequest.getActualReadStart() - readRequest.getBackendReadStart());
+            int suffixBufferLength = (int) (readRequest.getBackendReadEnd() - readRequest.getActualReadEnd());
+            log.debug(String.format("PrefixLength: %d SuffixLength: %d", prefixBufferLength, suffixBufferLength));
+
+            if (prefixBufferLength > 0) {
+                log.debug(String.format("Trying to Read %d bytes into prefix buffer", prefixBufferLength));
+                totalPrefixRead += readIntoBuffer(affixBuffer, 0, prefixBufferLength);
+                log.debug(String.format("Read %d bytes into prefix buffer", prefixBufferLength));
+                copyIntoCache(affixBuffer, 0, totalPrefixRead, readRequest.backendReadStart);
```

Comment:
though this will work fine but this conveys a wrong meaning. Change `copyIntoCache(affixBuffer, 0, totalPrefixRead, readRequest.backendReadStart);` to `copyIntoCache(affixBuffer, 0, prefixBufferLength, readRequest.backendReadStart);`